drivers/dai/intel/uaol/uaol.c

```diff
@@ -0,0 +1,1328 @@
+		min_length = sizeof(tlv->value.xhci_controller_bdf);
+		break;
+	case INTEL_UAOL_AUX_UAOL_CONFIG:
+		min_length = sizeof(tlv->value.uaol_config);
+		break;
+	case INTEL_UAOL_AUX_FIFO_SAO:
+		min_length = sizeof(tlv->value.uaol_fifo_sao);
+		break;
+	case INTEL_UAOL_AUX_USB_EP_INFO:
+	case INTEL_UAOL_AUX_USB_EP_FEEDBACK_INFO:
+		min_length = sizeof(tlv->value.usb_ep_info);
+		break;
+	case INTEL_UAOL_AUX_USB_ART_DIVIDER:
+		min_length = sizeof(tlv->value.art_divider);
+		break;
+	default:
+		return NULL;
+	}
+
+	if (tlv->length < min_length) {
+		return NULL;
+	}
+
+	*buf_pos += offsetof(struct intel_uaol_aux_tlv, value) + tlv->length;
+
+	return tlv;
+}
+
+static int uaol_set_uaol_config(struct dai_intel_uaol *dp, struct intel_uaol_config *cfg)
+{
+	if (dp->link_idx != cfg->link_idx) {
+		return -EINVAL;
+	}
+
+	if (dp->stream_idx == cfg->stream_idx) {
+		dp->stream_type = INTEL_UAOL_STREAM_TYPE_NORMAL;
+		dp->service_interval = INTEL_UAOL_SERVICE_INTERVAL_US;
+		return 0;
+	}
+
+	if (dp->stream_idx == cfg->feedback_idx) {
+		dp->stream_type = INTEL_UAOL_STREAM_TYPE_FEEDBACK;
+		dp->service_interval = uaol_convert_service_interval_4bit_to_us(
+						cfg->feedback_period);
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+int uaol_set_ep_info(struct dai_intel_uaol *dp, struct intel_uaol_usb_ep_info ep_info)
+{
+	if (dp->stream_type == INTEL_UAOL_STREAM_TYPE_NORMAL) {
+		dp->ep_info = ep_info;
+	}
+
+	return 0;
+}
+
+int uaol_set_ep_feedback_info(struct dai_intel_uaol *dp, struct intel_uaol_usb_ep_info ep_info)
+{
+	if (dp->stream_type == INTEL_UAOL_STREAM_TYPE_FEEDBACK) {
+		dp->ep_info = ep_info;
+	}
+
+	return 0;
+}
+
+static int uaol_set_aux_config(struct dai_intel_uaol *dp, const void *data, size_t data_size)
+{
+	size_t data_offset = 0;
+	struct intel_uaol_aux_tlv *tlv;
+	int ret = 0;
+
+	while (1) {
+		tlv = uaol_get_aux_tlv((uint8_t *)data, data_size, &data_offset);
+		if (!tlv) {
+			break;
+		}
+
+		switch (tlv->type) {
+		case INTEL_UAOL_AUX_XHCI_CONTROLLER_BDF:
+			dp->link->bdf = tlv->value.xhci_controller_bdf;
+			break;
+		case INTEL_UAOL_AUX_UAOL_CONFIG:
+			ret = uaol_set_uaol_config(dp, &tlv->value.uaol_config);
+			break;
+		case INTEL_UAOL_AUX_FIFO_SAO:
+			dp->link->fifo_sao = tlv->value.uaol_fifo_sao;
+			break;
+		case INTEL_UAOL_AUX_USB_EP_INFO:
+			ret = uaol_set_ep_info(dp, tlv->value.usb_ep_info);
+			break;
+		case INTEL_UAOL_AUX_USB_EP_FEEDBACK_INFO:
+			ret = uaol_set_ep_feedback_info(dp, tlv->value.usb_ep_info);
+			break;
+		case INTEL_UAOL_AUX_USB_ART_DIVIDER:
+			dp->link->art_divider = tlv->value.art_divider;
+			break;
+		default:
+			break;
+		}
+
+		if (ret) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * Gets a valid ioctl-type TLV from the buffer at the specified buffer position.
+ * On success, the buffer position is updated to point to the next possible TLV.
+ */
+static struct intel_uaol_ioctl_tlv *uaol_get_ioctl_tlv(uint8_t *buf, size_t buf_size,
+						       size_t *buf_pos)
```

Comment:
This appears to be implementing a wire protocol.  That's a pretty weird thing to be happening at the DAI level.   Doesn't this belong in the app somewhere, which translates to the C API exposed by the driver?

It seems like there's some layers of abstraction needed here.  Where is this protocol defined?  Who else implements it (I'm guessing something in the host kernel)?

Reply:
Thanks a lot for the comments! This code implements the IPC4_DMA_CONTROL protocol which is actually based on the series of TLV and needs some kind of parsing logic. Processing this at DAI level seemed reasonable to me (at least when it comes to Intel DAIs) as Intel DAIs are apparently expected to handle with IPC4 directly. I would like to stay with TLV processing at DAI level but with no ioctl extension for DAI interface.